### DISCUSSION TITLE
Feature Request: adding "when" clause to "yield return"

### DISCUSSION THREAD

**Discussion by jcouv** [dotnet/csharplang#1017] [General]
@mspost commented on [Wed Oct 18 2017](https://github.com/dotnet/roslyn/issues/22762)

Instead of writing this:
 
```
     IEnumerable<string> foo()
     {
           var stringValue = myVariable.GetOtherValue();
           if (stringValue != null)
                yield return stringValue;
     }

``` 
I would like to be able to write this:
 
     IEnumerable<string> foo()
     {
           yield return myVariable.GetOtherValue() when v => v != null;
     }
 
This seems like a natural extension to the way 'when' is used by exceptions.  Alternatively, given the way a property setter has a fixed value 'value', it could be done that way as well, like this:

```
     IEnumerable<string> foo()
     {
           yield return myVariable.GetOtherValue() when value != null;
     }

```

Since a property setter would never need a yield return, this may be a cleaner way to proceed.

---

@jcouv commented on [Wed Oct 18 2017](https://github.com/dotnet/roslyn/issues/22762#issuecomment-337792929)

This is a language design discussion, not a compiler bug. I'll move to csharplang repo.

**Comment by mspost** [comment]
Sorry.  This was my first post and I got the wrong location.

**Comment by MkazemAkhgary** [comment]
use of value can bring ambiguity between local variables.

     var value = 0;
     yield return myVariable.GetOtherValue() when value != null;

this issue could be supported by #973 and #867 

    (var value = myVariable.GetOtherValue()) != null ? yield return value : yield continue;

yield continue is another proposal in this comment :))

**Comment by MkazemAkhgary** [comment]
with when clause it would be like

    yield return (var result = myVariable.GetOtherValue()) when result != null;

when clause can access local and field variables, just like how its works after catch statements.

**Comment by orthoxerox** [comment]
I think declaration expressions will solve this problem for many language constructs instead of just `yield return`:

```c#
     IEnumerable<string> Foo()
     {
           if ((var stringValue = myVariable.GetOtherValue()) != null)
                yield return stringValue;
     }
```

**Comment by yaakov-h** [comment]
```csharp
if (myVariable.GetOtherValue() is {} val) yield return val;
```

**Comment by jnm2** [comment]
We've already had discussions on why `if` is preferable to appending a condition to a return statement.
See #325, #453 and #724

**Comment by Thaina** [comment]
I would prefer condition expression

```C#
IEnumerable<string> Foo()
{
    (myVariable?.GetOtherValue() is var v && v != null) ? yield return v : yield break;
}
// or maybe
IEnumerable<string> Foo()
{
    yield (myVariable?.GetOtherValue() is var v && v != null) ? return v : break;
}
```

**Comment by MgSam** [comment]
@yaakov-h 
``` C#
if (myVariable.GetOtherValue() is {} val) yield return val;
```
Yuck, that is awful. Shocked that this got 5 upvotes. Completely non-obvious that the intention here is to do a null check. You save one line of code for the potential of hours of bugs later on when someone comes later, doesn't understand your intent, and refactors to something that misses the null check.

**Comment by MkazemAkhgary** [comment]
@yaakov-h how does that even compile?

**Comment by Joe4evr** [comment]
@MkazemAkhgary [It's a proposed part of pattern matching.](https://github.com/dotnet/csharplang/issues/306#issuecomment-335948566)

**Comment by yaakov-h** [comment]
@MgSam I think it was upvoted not because it's a brilliant way to do this, but because it simply achieves the goals of the original post using C# language features that already exist and/or are actively being championed.

If you're concerned about refactoring breaking the code... write a unit test. 😉

**Comment by Thaina** [comment]
If it to be the same as pattern matching it should be `return when _expression_ : obj;`
